File: legacy/src/main/rs/emulate/legacy/config/ConfigPropertyType.java
package rs.emulate.legacy.config;

import rs.emulate.shared.property.PropertyType;

/**
 * A {@link PropertyType} used as part of the {@link ConfigEncoder} and {@link ConfigDecoder}, {@link #opcode()}. Should
 * <strong>only</strong> be implemented by enumerators (excluding the existing {@link DynamicConfigPropertyType} class).
 * 
 * @author Major
 */
public interface ConfigPropertyType extends PropertyType {

	/**
	 * Gets the opcode of this ConfigPropertyType.
	 * 
	 * @return The opcode.
	 */
	public int opcode();

	/**
	 * Gets the {@link SerializableProperty} associated with this ConfigPropertyType from the specified
	 * {@link MutableConfigDefinition}.
	 * 
	 * @param definition The MutableDefinition.
	 * @return The DefinitionProperty.
	 */
	default public <T> SerializableProperty<T> propertyFrom(MutableConfigDefinition definition) {
		return definition.getProperty(this);
	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/config/ConfigPropertyType.kt
package rs.emulate.legacy.config

import rs.emulate.shared.property.PropertyType

/**
 * A [PropertyType] used by the [ConfigEncoder] and [ConfigDecoder].
 */
abstract class ConfigPropertyType<T> : PropertyType {

    /**
     * The opcode of this ConfigPropertyType.
     * */
    abstract val opcode: Int

    override val name = javaClass.simpleName!!

    /**
     * Gets the [SerializableProperty] associated with this ConfigPropertyType from the specified
     * [MutableConfigDefinition].
     */
    fun propertyFrom(definition: MutableConfigDefinition): SerializableProperty<T> {
        return definition.getProperty(this)
    }

    override fun formattedName(): String {
        return javaClass.simpleName!!.replace("(.)([A-Z])".toRegex(), "$1 $2")
    }

}
